lib/core/test/unit/Cardano/Wallet/Primitive/AddressDiscovery/DelegationSpec.hs

```diff
@@ -0,0 +1,556 @@
+newtype StakeKey' (depth :: Depth) key = StakeKey' Word
+    deriving newtype (Eq, Enum, Ord, Show, Bounded)
+
+type StakeKey = StakeKey' 'AddressK XPub
+
+instance ToRewardAccount StakeKey' where
+    toRewardAccount (StakeKey' i) = RewardAccount . B8.pack $ show i
+    someRewardAccount = error "someRewardAccount: not implemented"
+
+instance HardDerivation StakeKey' where
+    type AddressIndexDerivationType StakeKey' = 'Soft
+    deriveAccountPrivateKey _ _ _ =
+        error "deriveAccountPrivateKey: not implemented"
+    deriveAddressPrivateKey _ _ _ _ =
+        error "deriveAddressPrivateKey not implemented"
+
+instance SoftDerivation StakeKey' where
+    deriveAddressPublicKey _acc _role i = StakeKey' $ toEnum $ fromEnum i
+
+instance MkKeyFingerprint StakeKey' Address where
+    paymentKeyFingerprint (Address addr) = Right $ KeyFingerprint $ B8.drop 4 addr
+
+instance PaymentAddress 'Mainnet StakeKey' where
+    liftPaymentAddress (KeyFingerprint fp) = Address fp
+    paymentAddress k = Address $ "addr" <> unRewardAccount (toRewardAccount k)
+
+instance MkKeyFingerprint StakeKey' (StakeKey' 'AddressK XPub) where
+    paymentKeyFingerprint k =
+        Right $ KeyFingerprint $ unRewardAccount (toRewardAccount k)
+
+--
+-- Mock chain of delegation certificates
+--
+
+instance Arbitrary RewardAccount where
+    arbitrary = toRewardAccount @StakeKey' <$> arbitrary
+
+instance Arbitrary Cert where
+    arbitrary = genericArbitrary
+    shrink = genericShrink
+
+instance Arbitrary (StakeKey' depth key) where
+    arbitrary = StakeKey' <$> arbitrary
+
+--
+-- Mock chain data
+--
+
+data Cmd
+    = CmdSetPortfolioOf Int
+    -- ^ Calls @setPortfolioOf@ which registers or de-registers keys to reach
+    -- the new target.
+    --
+    -- If the target is already met, the command has no effect.
+    --
+    -- Delegation certificates are /not/ generated for existing keys.
+    --
+    -- TODO: Also test arbitrary re-delegations.
+    | CmdOldWalletToggleFirstKey
+
+      -- ^ A wallet implementation without multi-stake-key support could decide
+      -- to either
+      -- 1. register stake-key 0 witout adding a pointer UTxO
+      -- 2. de-register stake-key 0 despite despite e.g. key 1 being active
+      -- depending on whether it is registered or not.
+      --
+      -- Having a "toggle"-command instead of two separate commands, makes
+      -- generating valid arbitrary values easier.
+    | CmdAdversarialReg RewardAccount
+      -- ^ Someone could pay 2 ada to (re-)register your stake key. Your wallet
+      -- shouldn't be affected negatively from it.
+    | CmdMimicPointerOutput RewardAccount
+      -- ^ Someone could send funds to the same UTxO
+    deriving (Generic, Eq)
+
+isAdversarial :: Cmd -> Bool
+isAdversarial (CmdSetPortfolioOf _) = False
+isAdversarial (CmdAdversarialReg _) = True
+isAdversarial CmdOldWalletToggleFirstKey = False
+isAdversarial (CmdMimicPointerOutput _) = True -- Kind of
```

Comment:
Don't be fooled by appearances; because they send us money, doesn't mean they are friendly :smile:, that's definitely adversarial.